src/mlpack/methods/ann/layer/atrous_convolution.hpp

```diff
@@ -166,45 +166,106 @@ class AtrousConvolution
                 arma::Mat<eT>&& gradient);
 
   //! Get the parameters.
-  OutputDataType const& Parameters() const { return weights; }
+  const OutputDataType& Parameters() const { return weights; }
   //! Modify the parameters.
   OutputDataType& Parameters() { return weights; }
 
   //! Get the output parameter.
-  OutputDataType const& OutputParameter() const { return outputParameter; }
+  const OutputDataType& OutputParameter() const { return outputParameter; }
   //! Modify the output parameter.
   OutputDataType& OutputParameter() { return outputParameter; }
 
   //! Get the delta.
-  OutputDataType const& Delta() const { return delta; }
+  const OutputDataType& Delta() const { return delta; }
   //! Modify the delta.
   OutputDataType& Delta() { return delta; }
 
   //! Get the gradient.
-  OutputDataType const& Gradient() const { return gradient; }
+  const OutputDataType& Gradient() const { return gradient; }
   //! Modify the gradient.
   OutputDataType& Gradient() { return gradient; }
 
   //! Get the input width.
-  size_t const& InputWidth() const { return inputWidth; }
+  const size_t& InputWidth() const { return inputWidth; }
   //! Modify input the width.
   size_t& InputWidth() { return inputWidth; }
 
   //! Get the input height.
-  size_t const& InputHeight() const { return inputHeight; }
+  const size_t& InputHeight() const { return inputHeight; }
   //! Modify the input height.
   size_t& InputHeight() { return inputHeight; }
 
   //! Get the output width.
-  size_t const& OutputWidth() const { return outputWidth; }
+  const size_t& OutputWidth() const { return outputWidth; }
   //! Modify the output width.
   size_t& OutputWidth() { return outputWidth; }
 
   //! Get the output height.
-  size_t const& OutputHeight() const { return outputHeight; }
+  const size_t& OutputHeight() const { return outputHeight; }
   //! Modify the output height.
   size_t& OutputHeight() { return outputHeight; }
 
+  //! Get the input size
+  const size_t& InputSize() const { return inSize; }
+
+  //! Get the output size.
+  const size_t& OutputSize() const { return outSize; }
+
+  //! Get the kernel size (width and height).
+  const std::tuple<size_t, size_t> KernelSize() const
+  {
+    return std::forward_as_tuple(kW, kH);
+  }
+  //! Modify the kernel size (width and height).
+  std::tuple<size_t&, size_t&> KernelSize()
```

Comment:
You are right, this could cause some confusions if used after training was started; but we could handle that as well, maybe we could integrate that into a new issue.

Reply:
In many places we do allow "internal" parameters to be modified for "expert" users.  In this case, so long as the parameters are modified before the network is reset, I think it should be fine.  In fact this makes configuration straightforward, as you can do things like...

```
AtrousConvolution layer;
std::get<0>(layer.KernelSize()) = 10;
std::get<1>(layer.KernelSize()) = 5;
```

In this way we can avoid the need for passing a ton of parameters to the constructor.  However, it's possible in my example there I overlooked something, so let me know if I did.  I'm happy to figure out ways to make this safer, if you have any ideas.

Reply:
Yes, we can couple this with the above issue. I think we can fix it. :+1:

Reply:
I agree with @rcurtin here because this extra flexibility might be needed for some typical cases. However, I'm not sure how the other frameworks handle this, maybe I can take a look and get back at this.

Reply:
Keras, I see, doesn't allow straightforward re-setting of layer dimensions but I think it shouldn't be much of a harm if we keep it here.